File: app/src/main/java/com/pluu/webtoon/ui/WebtoonListFragment.java
package com.pluu.webtoon.ui;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.ContextCompat;
import android.support.v7.graphics.Palette;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.Toast;

import com.pluu.event.RxBusProvider;
import com.pluu.support.impl.AbstractWeekApi;
import com.pluu.support.impl.NAV_ITEM;
import com.pluu.support.impl.ServiceConst;
import com.pluu.webtoon.AppController;
import com.pluu.webtoon.R;
import com.pluu.webtoon.adapter.MainListAdapter;
import com.pluu.webtoon.common.Const;
import com.pluu.webtoon.db.RealmHelper;
import com.pluu.webtoon.event.MainEpisodeLoadedEvent;
import com.pluu.webtoon.event.MainEpisodeStartEvent;
import com.pluu.webtoon.item.WebToonInfo;
import com.pluu.webtoon.utils.GlideUtils;

import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;


/**
 * Main EpisodePage List Fragment
 * Created by PLUUSYSTEM-NEW on 2015-10-27.
 */
public class WebtoonListFragment extends Fragment {
    private final String TAG = WebtoonListFragment.class.getSimpleName();

    @Inject
    RealmHelper realmHelper;

    private RecyclerView recyclerView;
    private GridLayoutManager manager;
    private int position;

    private final int REQUEST_DETAIL = 1000;
    public static final int REQUEST_DETAIL_REFERRER = 1001;

    private AbstractWeekApi serviceApi;
    private int columnCount;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((AppController) getContext().getApplicationContext()).getRealmHelperComponent().inject(this);

        NAV_ITEM service = ServiceConst.getApiType(getArguments());
        serviceApi = AbstractWeekApi.getApi(getContext(), service);
        position = getArguments().getInt(Const.EXTRA_POS);
        columnCount = getResources().getInteger(R.integer.webtoon_column_count);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        recyclerView = (RecyclerView) inflater.inflate(R.layout.fragment_webtoon_list, container, false);
        manager = new GridLayoutManager(getActivity(), columnCount);
        recyclerView.setLayoutManager(manager);

        return recyclerView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        getApiRequest()
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .map(getFavoriteProcessFunc())
            .map(unsortedList -> {
                // 정렬
                Collections.sort(unsortedList);
                return unsortedList;
            })
            .doOnSubscribe(disposable -> RxBusProvider.getInstance().send(new MainEpisodeStartEvent()))
            .doOnSuccess(items -> RxBusProvider.getInstance().send(new MainEpisodeLoadedEvent()))
            .subscribe(getRequestSubscriber(), throwable -> {
                RxBusProvider.getInstance().send(new MainEpisodeLoadedEvent());
                Toast.makeText(getActivity(), throwable.getMessage(), Toast.LENGTH_SHORT).show();
            });
    }

    //	@RxLogSubscriber
    @NonNull
    private Consumer<List<WebToonInfo>> getRequestSubscriber() {
        return list -> {
            final FragmentActivity activity = getActivity();
            if (activity == null || activity.isFinishing()) {
                return;
            }

            recyclerView.setAdapter(new MainListAdapter(activity, list) {
                @Override
                public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
                    ViewHolder vh = super.onCreateViewHolder(viewGroup, i);
                    setClickListener(vh);
                    return vh;
                }
            });
        };
    }

    @NonNull
    private Function<List<WebToonInfo>, List<WebToonInfo>> getFavoriteProcessFunc() {
        return list -> {
            for (WebToonInfo item : list) {
                item.setIsFavorite(
                    realmHelper.getFavoriteToon(serviceApi.getNaviItem(), item.getToonId())
                );
            }
            return list;
        };
    }

    //	@RxLogObservable
    private Single<List<WebToonInfo>> getApiRequest() {
        return Single.fromCallable(() -> serviceApi.parseMain(position));
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode != Activity.RESULT_OK) {
            return;
        }

        if (requestCode == REQUEST_DETAIL) {
            // 즐겨찾기 변경 처리 > 다른 ViewPager의 Fragment도 수신받기위해 Referrer
            Fragment frag = getFragmentManager().findFragmentByTag(Const.MAIN_FRAG_TAG);
            if (frag != null) {
                frag.onActivityResult(REQUEST_DETAIL_REFERRER, resultCode, data);
            }
        } else if (requestCode == REQUEST_DETAIL_REFERRER) {
            // ViewPager 로부터 전달받은 Referrer
            WebToonInfo info = data.getParcelableExtra(Const.EXTRA_EPISODE);
            favoriteUpdate(info);
        }
    }

    private void favoriteUpdate(WebToonInfo info) {
        MainListAdapter adapter = (MainListAdapter) recyclerView.getAdapter();
        int position = adapter.modifyInfo(info);
        if (position != -1) {
            adapter.notifyItemChanged(position);
        }
    }

    private void setClickListener(final MainListAdapter.ViewHolder vh) {
        View v = vh.itemView;
        v.setOnClickListener(v1 -> {
            final WebToonInfo item = (WebToonInfo) vh.titleView.getTag();
            if (item.isLock()) {
                Toast.makeText(getContext(),
                    R.string.msg_not_support,
                    Toast.LENGTH_SHORT).show();
            } else {
                loadPalette(vh.thumbnailView, item);
            }
        });
    }

    private void loadPalette(ImageView view, final WebToonInfo item) {
        Bitmap bitmap = GlideUtils.loadGlideBitmap(view);
        if (bitmap != null) {
            asyncPalette(item, bitmap);
        }
    }

    private void asyncPalette(final WebToonInfo item, Bitmap bitmap) {
        final Context context = getActivity();
        Palette.from(bitmap).generate(p -> {
            int bgColor = p.getDarkVibrantColor(
                Color.BLACK);
            int statusColor = p.getDarkMutedColor(
                ContextCompat.getColor(context, R.color.theme_primary_dark));
            moveEpisode(item, bgColor, statusColor);
        });
    }

    private void moveEpisode(WebToonInfo item, int bgColor, int statusColor) {
        Intent intent = new Intent(getActivity(), EpisodesActivity.class);
        intent.putExtra(Const.EXTRA_API, serviceApi.getNaviItem());
        intent.putExtra(Const.EXTRA_EPISODE, item);
        intent.putExtra(Const.EXTRA_MAIN_COLOR, bgColor);
        intent.putExtra(Const.EXTRA_STATUS_COLOR, statusColor);
        startActivityForResult(intent, REQUEST_DETAIL);
    }

    private void updateSpanCount() {
        int columnCount = getResources().getInteger(R.integer.webtoon_column_count);
        manager.setSpanCount(columnCount);
        recyclerView.getAdapter().notifyDataSetChanged();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);

        if (newConfig.orientation == Configuration.ORIENTATION_LANDSCAPE
            || newConfig.orientation == Configuration.ORIENTATION_PORTRAIT) {
            updateSpanCount();
        }
    }
}

File: app/src/main/java/com/pluu/webtoon/ui/WebtoonListFragment.kt
package com.pluu.webtoon.ui

import android.app.Activity
import android.content.Intent
import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.palette.graphics.Palette
import androidx.recyclerview.widget.GridLayoutManager
import com.bumptech.glide.load.resource.gif.GifDrawable
import com.pluu.event.RxBusProvider
import com.pluu.kotlin.getCompatColor
import com.pluu.kotlin.toVisibleOrGone
import com.pluu.kotlin.toast
import com.pluu.support.impl.AbstractWeekApi
import com.pluu.support.impl.ServiceConst
import com.pluu.webtoon.R
import com.pluu.webtoon.adapter.MainListAdapter
import com.pluu.webtoon.common.Const
import com.pluu.webtoon.db.RealmHelper
import com.pluu.webtoon.di.Properties
import com.pluu.webtoon.event.MainEpisodeLoadedEvent
import com.pluu.webtoon.event.MainEpisodeStartEvent
import com.pluu.webtoon.item.WebToonInfo
import com.pluu.webtoon.ui.listener.WebToonSelectListener
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer
import io.reactivex.functions.Function
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_webtoon_list.*
import org.koin.android.ext.android.inject
import org.koin.android.ext.android.setProperty

/**
 * Main EpisodePage List Fragment
 * Created by pluu on 2017-05-07.
 */
class WebtoonListFragment : Fragment(), WebToonSelectListener {

    private val realmHelper: RealmHelper by inject()

    private val serviceApi: AbstractWeekApi by inject(Properties.WEEK_KEY)

    private var position: Int = 0
    private val REQUEST_DETAIL = 1000

    private val disposables: CompositeDisposable by lazy {
        CompositeDisposable()
    }

    private val toonLayoutManager: GridLayoutManager by lazy {
        GridLayoutManager(activity, resources.getInteger(R.integer.webtoon_column_count))
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_webtoon_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setProperty(Properties.API_TYPE_KEY, ServiceConst.getApiType(arguments))
        position = arguments?.getInt(Const.EXTRA_POS) ?: 0

        with(recyclerView) {
            layoutManager = toonLayoutManager
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        Single.fromCallable { serviceApi.parseMain(position) }
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .map(favoriteProcessFunc)
            .map { unsortedList ->
                // 정렬
                unsortedList.sorted()
            }
            .doOnSubscribe { RxBusProvider.getInstance().send(MainEpisodeStartEvent()) }
            .doOnSuccess { RxBusProvider.getInstance().send(MainEpisodeLoadedEvent()) }
            .subscribe(requestSubscriber, Consumer { t ->
                RxBusProvider.getInstance().send(MainEpisodeLoadedEvent())
                Toast.makeText(activity, t.message, Toast.LENGTH_SHORT).show()
            }).let {
                disposables.add(it)
            }
    }

    override fun onDestroyView() {
        disposables.clear()
        super.onDestroyView()
    }

    private val requestSubscriber = Consumer<List<WebToonInfo>> { list ->
        val activity = activity?.takeUnless { it.isFinishing } ?: return@Consumer
        recyclerView.adapter = MainListAdapter(activity, list, this)
        emptyView.visibility = list.isEmpty().toVisibleOrGone()
    }

    private val favoriteProcessFunc = Function<List<WebToonInfo>, List<WebToonInfo>> { list ->
        list.forEach {
            it.isFavorite = realmHelper.getFavoriteToon(serviceApi.naviItem, it.toonId)
        }
        list
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != Activity.RESULT_OK) {
            return
        }

        if (requestCode == REQUEST_DETAIL) {
            // 즐겨찾기 변경 처리 > 다른 ViewPager의 Fragment도 수신받기위해 Referrer
            fragmentManager?.findFragmentByTag(Const.MAIN_FRAG_TAG)
                ?.onActivityResult(REQUEST_DETAIL_REFERRER, resultCode, data)
        } else if (requestCode == REQUEST_DETAIL_REFERRER) {
            // ViewPager 로부터 전달받은 Referrer
            data?.getParcelableExtra<WebToonInfo>(Const.EXTRA_EPISODE)?.apply {
                favoriteUpdate(this)
            }
        }
    }

    private fun favoriteUpdate(info: WebToonInfo) {
        val adapter = recyclerView.adapter as MainListAdapter
        val position = adapter.modifyInfo(info)
        if (position != -1) {
            adapter.notifyItemChanged(position)
        }
    }

    private fun loadPalette(view: ImageView, item: WebToonInfo) {
        view.palletBitmap?.let {
            asyncPalette(item, it)
        }
    }

    private fun asyncPalette(item: WebToonInfo, bitmap: Bitmap) {
        val context = context ?: return
        Palette.from(bitmap).generate { p ->
            val bgColor = p?.getDarkVibrantColor(Color.BLACK) ?: Color.BLACK
            val statusColor =
                p?.getDarkMutedColor(context.getCompatColor(R.color.theme_primary_dark))
                    ?: context.getCompatColor(R.color.theme_primary_dark)
            moveEpisode(item, bgColor, statusColor)
        }
    }

    private fun moveEpisode(item: WebToonInfo, bgColor: Int, statusColor: Int) {
        startActivityForResult(Intent(activity, EpisodesActivity::class.java).apply {
            putExtra(Const.EXTRA_API, serviceApi.naviItem)
            putExtra(Const.EXTRA_EPISODE, item)
            putExtra(Const.EXTRA_MAIN_COLOR, bgColor)
            putExtra(Const.EXTRA_STATUS_COLOR, statusColor)
        }, REQUEST_DETAIL)
    }

    private fun updateSpanCount() {
        toonLayoutManager.spanCount = resources.getInteger(R.integer.webtoon_column_count)
        recyclerView.adapter?.notifyDataSetChanged()
    }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        super.onConfigurationChanged(newConfig)

        newConfig?.takeIf {
            it.orientation == Configuration.ORIENTATION_LANDSCAPE
                    || it.orientation == Configuration.ORIENTATION_PORTRAIT
        }.run {
            updateSpanCount()
        }
    }

    override fun selectLockItem() {
        toast(R.string.msg_not_support)
    }

    override fun selectSuccess(view: ImageView, item: WebToonInfo) {
        loadPalette(view, item)
    }

    companion object {
        const val REQUEST_DETAIL_REFERRER = 1001
    }
}

private val ImageView.palletBitmap: Bitmap?
    get() {
        val innerDrawable: Drawable = drawable
        return when (innerDrawable) {
            is BitmapDrawable -> innerDrawable.bitmap
            is GifDrawable -> innerDrawable.firstFrame
            else -> null
        }
    }